File: 프로그래머스/3/161988. 연속 펄스 부분 수열의 합/연속 펄스 부분 수열의 합.js
function solution(sequence) {
    var dp1 = sequence.map((v, i)=>(
        (i % 2 == 0 ? 1 : -1) * v
    ))
    
    var dp2 = sequence.map((v, i)=>(
        (i % 2 != 0 ? 1 : -1) * v
    ))
    
    for(var i = 1; i < sequence.length; i++){
        
        dp1[i] = Math.max(dp1[i-1] + dp1[i], dp1[i])
        dp2[i] = Math.max(dp2[i-1] + dp2[i], dp2[i])
    }
    
    dp1.sort((a,b)=>a - b)
    dp2.sort((a,b)=>a - b)
    
    return Math.max(dp1[dp1.length - 1], dp2[dp1.length - 1])
}

File: 프로그래머스/3/161988. 연속 펄스 부분 수열의 합/연속 펄스 부분 수열의 합.java
import java.util.stream.*;
import java.util.*;
class Solution {
    public long solution(int[] sequence) {
        long[] dp1 = new long[sequence.length];
        long[] dp2 = new long[sequence.length];
        for(int i = 0; i < sequence.length; i++){
            dp1[i] = (i % 2 == 0 ? 1 : -1) * sequence[i];
            dp2[i] = (i % 2 == 0 ? -1 : 1) * sequence[i];
        }
        
        for(int i = 1; i < sequence.length; i++){
            dp1[i] = Math.max(dp1[i - 1] + dp1[i], dp1[i]);
            dp2[i] = Math.max(dp2[i - 1] + dp2[i], dp2[i]);
        }
        long a = Arrays.stream(dp1).max().orElse(0);
        long b = Arrays.stream(dp2).max().orElse(0);
        return Math.max(a,b);
    }
}